Add PP-MiniLM

LiuChiachi opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/PaddleNLP/pull/26 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ Models | APIs | Docs | Others ] -->
Models & Docs
### Description
<!-- Describe what this PR does -->
- Add PP-MiniLM code
- Add doc for PP-MiniLM

Model | #Params | #FLOPs | Speedup | AFQMC | TNEWS | IFLYTEK | CMNLI | OCNLI | WSC | CSL | CLUE平均值
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Bertbase | 102.3M | 10.87B | 1.00x | 74.14 | 56.81 | 61.10 | 81.19 | 74.85 | 79.93 | 81.47 | 72.78
TinyBERT6 | 59.7M | 5.44B | 1.66x | 72.59 | 55.70 | 57.64 | 79.57 | 73.97 | 77.63 | 80.00 | 71.01
UER-py RoBERTa L6- H768 | 59.7M | 5.44B | 1.66x | 69.74 | 66.36 | 59.95 | 77.00 | 71.39 | 71.05 | 82.83 | 71.19
RBT6, Chinese | 59.7M | 5.44B | 1.66x | 73.93 | 56.63 | 59.79 | 79.28 | 73.12 | 77.30 | 80.80 | 71.55
ERNIE-Tiny | 90.7M | 4.83B | 1.89x | 70.67 | 55.60 | 59.91 | 75.74 | 71.36 | 67.11 | 76.70 | 68.16
PP-MiniLM 6L-768H | 59.7M | 5.44B | 1.66x | 74.14 | 57.43 | 61.75 | 81.01 | 76.17 | 86.18 | 77.47 | 73.45
PP-MiniLM裁剪后 | 49.1M | 4.08B | 2.00x | 73.91 | 57.44 | 61.64 | 81.10 | 75.59 | 85.86 | 77.97 | 73.36
PP-MiniLM量化后 | 49.2M | 4.08B | 4.15x | 74.00 | 57.37 | 61.33 | 81.09 | 75.56 | 85.85 | 76.53 | 73.10

TODO:
1.更新QA对UER-py的测试结果进README；在cuda10.2 paddle2.2.1下测试CSL；

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+PP-MiniLM融合了蒸馏、裁剪、量化、高性能推理技术，拥有精度高、推理速度快、参数规模小的特点：
+
+- 精度高：6层-768hidden size的模型，精度高于华为、腾讯同样大小的模型；
```
不要用公司名，要用模型名。

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+# PP-MiniLM中文特色小模型
+
+
+PP-MiniLM中文特色小模型，模型结构同ERNIE，目前本案例主要包含六层transformer layer的模型的通用蒸馏，以及借助PaddleSlim对模型进行裁剪和量化，进一步提升推理速度。
```
模型介绍中需要是否突出下基于MiniLMv2策略的改进？@tianxin1860

tianxin1860 replied:
如线上沟通，按照 1.推理速度快 2.模型效果好 3.参数规模小 的逻辑来呈现。在`模型效果好`小项里体现我们对 MiniLMv2 的改进。

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| -------------------- | ------------- | ------ | ------- | ----- | ----- | ------- | ----- | ----- | ----- | ----- | ---------- |
+| bert-base-chinese    | 102.27M       |        | TODO    |       |       |         |       |       |       |       |            |
+| TinyBERT(6l-768d)    | 59.7M         |        | 1.00x   | 72.22 | 55.82 | 58.10   | 79.53 | 74.00 | 75.99 | 80.57 | 70.89      |
+| 腾讯 RoBERTa 6l-768d | 59.7M         |        | 1.00x   | 69.74 | 66.36 | 59.95   | 77.00 | 71.39 | 71.05 | 82.83 | 71.19      |
```
UER-py RoBERTa xxxx  去掉公司名

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+### 数据介绍
+
+百度内部业务数据。数据被分割成64个文件，放在目录dataset下。
```
去掉内部业务数据这段话。是否改用CLUESmall数据来作为数据示例。

ZeyuChen replied:
或者只是简单说下数据的组织方式即可

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+PP-MiniLM模型的蒸馏方法介绍：
+
+用large-size教师模型的第20层对6层学生模型第6层的q与q、k与k、v与v之间的样本间关系进行蒸馏。即对q、k、v统一head_num之后进行重新排列，
```
large-size教师模型是什么？是否需要以某个模型为例？

ZeyuChen replied:
这段话需要重新概括

ZeyuChen replied:
不适特别清晰

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+|         | 平均耗时(s) | 加速比  |
+| ------- | ----------- | ------- |
+| float32 | 12.6        | 0       |
```
FP32

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| ------- | ----------- | ------- |
+| float32 | 12.6        | 0       |
+| 裁剪后   | 10.98       | 14.84%  |
+| int8    | 5.8         | 117.59% |
```
Int8

ZeyuChen commented on examples/model_compression/PP-MiniLM/finetuning/run_all_search.sh:
```diff
@@ -0,0 +1,36 @@
+
```
删除空行

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| -------------------- | ------------- | ------ | ------- | ----- | ----- | ------- | ----- | ----- | ----- | ----- | ---------- |
+| bert-base-chinese    | 102.27M       |        | TODO    |       |       |         |       |       |       |       |            |
+| TinyBERT(6l-768d)    | 59.7M         |        | 1.00x   | 72.22 | 55.82 | 58.10   | 79.53 | 74.00 | 75.99 | 80.57 | 70.89      |
+| 腾讯 RoBERTa 6l-768d | 59.7M         |        | 1.00x   | 69.74 | 66.36 | 59.95   | 77.00 | 71.39 | 71.05 | 82.83 | 71.19      |
```
改用UER-py RoBERTa xxxx  去掉公司名

ZeyuChen replied:
是否需要测试下RBT6模型？

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| bert-base-chinese    | 102.27M       |        | TODO    |       |       |         |       |       |       |       |            |
+| TinyBERT(6l-768d)    | 59.7M         |        | 1.00x   | 72.22 | 55.82 | 58.10   | 79.53 | 74.00 | 75.99 | 80.57 | 70.89      |
+| 腾讯 RoBERTa 6l-768d | 59.7M         |        | 1.00x   | 69.74 | 66.36 | 59.95   | 77.00 | 71.39 | 71.05 | 82.83 | 71.19      |
+| PP-MiniLM 6l-768d    | 59.7M         |        | 1.00x   | 74.28 | 57.33 | 61.72   | 81.06 | 76.2  | 86.51 | 78.77 | 73.70      |
```
6l-768d -> 6L768H
使用大写，d改为H

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| PP-MiniLM 6l-768d    | 59.7M         |        | 1.00x   | 74.28 | 57.33 | 61.72   | 81.06 | 76.2  | 86.51 | 78.77 | 73.70      |
+| PP-MiniLM裁剪后      | 49.1M (+裁剪) |        | 1.15x   | 73.82 | 57.33 | 61.60   | 81.38 | 76.20 | 85.52 | 79.00 | 73.55      |
+| PP-MiniLM量化后      | 49.2M(+量化)  |        | 2.18x   | 73.61 | 57.18 | 61.49   | 81.26 | 76.31 | 84.54 | 77.67 | 73.15      |
+| ERNIE Tiny Subword   | 90.7M         |        | 1.08x   | 70.78 | 55.70 | 59.95   | 75.40 | 70.98 | 67.43 | 76.60 | 68.12      |
```
ERNIE-Tiny

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+整体效果一览：
+
+| model                | #Params       | #FLOPs | Speedup | AFQMC | TNEWS | IFLYTEK | CMNLI | OCNLI | WSC   | CSL   | CLUE平均值 |
```
model -> Model

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+| model                | #Params       | #FLOPs | Speedup | AFQMC | TNEWS | IFLYTEK | CMNLI | OCNLI | WSC   | CSL   | CLUE平均值 |
+| -------------------- | ------------- | ------ | ------- | ----- | ----- | ------- | ----- | ----- | ----- | ----- | ---------- |
+| bert-base-chinese    | 102.27M       |        | TODO    |       |       |         |       |       |       |       |            |
```
以BERT Base为1.x

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| ----- | ----- | ------- | ----- | ----- | ----- | ----- | ---------- |
+| 73.61 | 57.18 | 61.49   | 81.26 | 76.31 | 84.54 | 77.67 | 73.15      |
+
+## 利用PaddleInference进行预测部署
```
Paddle Inference，中间有空格

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+### 性能测试
+
+我们在NVIDIA 16G T4单卡上，使用inference/infer.py脚本，对量化后的模型进行预测。
```
NVIDIA Tesla T4  （T4只有16G，可以无需特别强调）

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
````diff
@@ -0,0 +1,297 @@
+```shell
+cd inference
+
+python infer.py --task_name ${task}  --model_path  ../quantization/${task}_quant_models/${algo}${bs}/int8  --int8 --use_trt --collect_shape # 生成shape range info文件
````
collect shape环节是否需要增强下说明

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+### 环境要求：
+
+这一步需要依赖paddle2.2.1，如果想要看到更明显的加速比，需要在T系列卡上测试（本案例使用的是T4）。若在V系列卡上测试，由于其不支持int8 tensor core，加速效果将达不到本文档表格中的加速效果。
```
要注重英文术语大小写拼写的正确性
Int8 Tensor Core

ZeyuChen replied:
如果需要得到更明显的加速效果，推荐在NVIDA Tensor Core GPU（如T4、A10、A100)上进行测试。

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
````diff
@@ -0,0 +1,297 @@
+float32预测脚本：
+
+```shell
+python infer.py --task_name ${task}  --model_path  $MODEL_PATH  --use_trt --collect_shape
````
Collect shape环节需要单独说明一下比较好，否则这里会让用户很困惑

ZeyuChen commented on examples/model_compression/PP-MiniLM/inference/infer.py:
```diff
@@ -0,0 +1,319 @@
+                config.tensorrt_engine_enabled()))
+            if args.collect_shape:
+                config.collect_shape_range_info(
+                    os.path.dirname(args.model_path) + "/" + args.task_name +
```
目录的拼写应该是用os.path.join API来拼接，强制用 /会导致windows不兼容

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+### 数据介绍
+
+本实验基于CLUE中分类数据集，linux系统下该数据集会在启动脚本后自动下载到`~/.paddlenlp/datasets/Clue/`。
```
本实验基于 [CLUE](url) 数据集，运行 Fine-tune 脚本会自动下载该数据集到 *** 目录.

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+本实验基于CLUE中分类数据集，linux系统下该数据集会在启动脚本后自动下载到`~/.paddlenlp/datasets/Clue/`。
+
+使用以下超参范围对第一步通用蒸馏得到的通用模型`GENERAL_MODEL_DIR`进行精调
```
基于如下超参范围对第一步蒸馏产出的小模型 `GENERAL_MODEL_DIR` 进行 Grid Search 超参寻优

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
````diff
@@ -0,0 +1,297 @@
+cd ofa
+```
+
+经过我们的实验，模型的宽度压缩为原来的3/4的情况下，模型精度无损（-0.15)。
````
注意表述: 6L768H 条件下，模型宽度压缩为原来的 3/4, 精度几乎无损。

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+经过我们的实验，模型的宽度压缩为原来的3/4的情况下，模型精度无损（-0.15)。
+
+### 压缩和蒸馏的启动脚本
```
是否明确一下 压缩、裁剪、蒸馏的关系以及使用场合？感觉标题这里用压缩和蒸馏可能引起误解。

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+    "cmnli": Accuracy,
+    "cluewsc2020": Accuracy,
+    "csl": Accuracy,
+    "xnli": Accuracy,
```
删除 CLUE 之外的数据集更合适一些？

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+    "roberta": (RobertaForSequenceClassification, RobertaTokenizer),
+    "ernie-gram": (ErnieGramForSequenceClassification, ErnieGramTokenizer),
+    "tinybert": (TinyBertForSequenceClassification, TinyBertTokenizer),
+    "nezha": (NeZhaForSequenceClassification, NeZhaTokenizer)
```
除 README 体现出来的模型之外，剩余模型的是否删除比较合适？

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+            'target']['span1_text'], example['target']['span2_text'], example[
+                'target']['span1_index'], example['target']['span2_index']
+        text_list = list(text)
+        # print(text)
```
多余的注释

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+            text_list.insert(query_idx + 2, "_")
+            text_list.insert(query_idx + len(query) + 2 + 1, "_")
+        text = "".join(text_list)
+        # print(text)
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+        tokenizer=tokenizer,
+        label_list=dev_ds.label_list,
+        max_seq_length=args.max_seq_length)
+    # train_ds = train_ds.map(trans_func, lazy=True)
+    # train_batch_sampler = paddle.io.DistributedBatchSampler(
+    # train_ds, batch_size=args.batch_size, shuffle=True)
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+        Pad(axis=0, pad_val=tokenizer.pad_token_type_id),  # segment
+        Stack(dtype="int64" if dev_ds.label_list else "float32")  # label
+    ): fn(samples)
+    # train_data_loader = DataLoader(
+    #     dataset=train_ds,
+    #     batch_sampler=train_batch_sampler,
+    #     collate_fn=batchify_fn,
+    #     num_workers=0,
+    #     return_list=True)
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+    for batch in dev_data_loader:
+        input_ids, segment_ids, labels = batch
+        logits = model(input_ids, segment_ids)
+        # loss = loss_fct(logits, labels)
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+    args.model_type = args.model_type.lower()
+    model_class, tokenizer_class = MODEL_CLASSES[args.model_type]
+
+    if args.task_name in ('xnli', 'chnsenticorp', 'lcqmc', 'dbqa'):
+        train_ds = load_dataset(args.task_name, splits='train')
+    else:
```
CLUE 数据集之外的相关逻辑建议删除

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.py:
```diff
@@ -0,0 +1,494 @@
+        num_workers=0,
+        return_list=True)
+
+    if args.task_name in ('xnli', 'chnsenticorp', 'lcqmc', 'dbqa'):
+        dev_ds = load_dataset(args.task_name, splits='dev')
+    else:
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/finetuning/run_clue.sh:
```diff
@@ -0,0 +1,20 @@
+export CUDA_VISIBLE_DEVICES=$6
```
$2、$3、$4 这种变量是否在脚本顶部先用变量解析出来更统一一些？

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer.py:
```diff
@@ -0,0 +1,319 @@
+            if len(data) == 2:
+                output = self.predict_batch(data)
+            else:
+                #ones = paddle.ones_like(data[0], dtype="int64")
+                #seq_length = paddle.cumsum(ones, axis=-1)
+                #position_ids = seq_length - ones
+                #attention_mask = paddle.unsqueeze((data[0] == paddle.to_tensor((0))).astype(paddle.get_default_dtype()) * -1e9, axis=[1, 2])
+                #output = self.predict_batch(data[: 2]+[position_ids, attention_mask])
```
多余的注释

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer.py:
```diff
@@ -0,0 +1,319 @@
+                #position_ids = seq_length - ones
+                #attention_mask = paddle.unsqueeze((data[0] == paddle.to_tensor((0))).astype(paddle.get_default_dtype()) * -1e9, axis=[1, 2])
+                #output = self.predict_batch(data[: 2]+[position_ids, attention_mask])
+                if i == 20:
```
这里的硬编码 i == 20 的意图是？

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer.py:
```diff
@@ -0,0 +1,319 @@
+            res = metric.accumulate()
+            print("task name: %s, acc: %s, " % (args.task_name, res), end='')
+        print("time: ", self.time)
+        #print(output)
```
多余注释

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer.sh:
```diff
@@ -0,0 +1,26 @@
+echo 原来的模型
```
脚本里统一使用英文？

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer.sh:
```diff
@@ -0,0 +1,26 @@
+python  infer.py  --task_name tnews --model_path  tnews/float   --use_trt
+
+
+echo 裁剪后
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer.sh:
```diff
@@ -0,0 +1,26 @@
+python infer.py --task_name tnews --model_path ofa_models/TNEWS/0.75/sub_static/float  --use_trt
+
+
+echo int8推理
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer_run.py:
```diff
@@ -0,0 +1,316 @@
+            if len(data) == 2:
+                output = self.predict_batch(data)
+            else:
+                #ones = paddle.ones_like(data[0], dtype="int64")
+                #seq_length = paddle.cumsum(ones, axis=-1)
+                #position_ids = seq_length - ones
+                #attention_mask = paddle.unsqueeze((data[0] == paddle.to_tensor((0))).astype(paddle.get_default_dtype()) * -1e9, axis=[1, 2])
+                #output = self.predict_batch(data[: 2]+[position_ids, attention_mask])
```
多余注释

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer_run.py:
```diff
@@ -0,0 +1,316 @@
+            res = metric.accumulate()
+            print("task name: %s, acc: %s, " % (args.task_name, res), end='')
+        print("time: ", self.time)
+        #print(output)
```
多余注释

tianxin1860 commented on examples/model_compression/PP-MiniLM/inference/infer_run.py:
```diff
@@ -0,0 +1,316 @@
+    model_class, tokenizer_class = MODEL_CLASSES[args.model_type]
+
+    dev_ds = load_dataset('clue', args.task_name, splits='dev')
+    tokenizer = tokenizer_class.from_pretrained(
+        '../general_distill/ernie-batchbatch-50w_400000/')
```
尽量避免 hard code

tianxin1860 commented on examples/model_compression/PP-MiniLM/ofa/export_model.py:
```diff
@@ -0,0 +1,226 @@
+import paddle.nn as nn
+import paddle.nn.functional as F
+
+from paddlenlp.transformers import BertModel, BertForSequenceClassification, BertTokenizer
```
BERTModel 会用到吗？

tianxin1860 commented on examples/model_compression/PP-MiniLM/ofa/export_model.py:
```diff
@@ -0,0 +1,226 @@
+        if isinstance(sublayer, paddle.nn.MultiHeadAttention):
+            sublayer.num_heads = int(args.width_mult * sublayer.num_heads)
+
+    #for name, params in origin_model.named_parameters():
+    #    print(name, params.name)
```
多余注释

tianxin1860 commented on examples/model_compression/PP-MiniLM/ofa/export_model.py:
```diff
@@ -0,0 +1,226 @@
+        os.makedirs(output_dir)
+    model_to_save = origin_model_new
+    model_to_save.save_pretrained(output_dir)
+    #print(origin_model_new.state_dict().keys())
+    #print("=====================")
+    #for name, params in origin_model_new.named_parameters():
+    #    print(name, params.name)
```
同上

tianxin1860 commented on examples/model_compression/PP-MiniLM/ofa/run_ofa.sh:
```diff
@@ -0,0 +1,19 @@
+export CUDA_VISIBLE_DEVICES=$6
+export TASK_NAME=$1
+export BATCH_SIZE=$3
+export SEQ_LEN=$5
+export PRE_EPOCHS=$4
+export LR=$2
+export STUDENT_DIR=$7
```
是否按 $1、$2 顺序依次解析变量？

tianxin1860 commented on examples/model_compression/PP-MiniLM/quantization/quant_all.sh:
```diff
@@ -0,0 +1,8 @@
+
+do
+
+    python quant_post.py --task_name ${task} --input_dir ${MODEL_DIR}/${task}/0.75/sub_static
```
这个 0.75 直接作为目录名吗？

tianxin1860 commented on examples/model_compression/PP-MiniLM/quantization/quant_post.py:
```diff
@@ -0,0 +1,176 @@
+            'target']['span1_text'], example['target']['span2_text'], example[
+                'target']['span1_index'], example['target']['span2_index']
+        text_list = list(text)
+        # print(text)
```
多余注释

tianxin1860 commented on paddlenlp/transformers/distill_utils.py:
```diff
@@ -20,9 +20,88 @@
+    s_head_dim, t_head_dim = s.shape[3], t.shape[3]
+
+    if alpha + beta == 1.0:
+        loss1 = 0.0
```
loss1、loss2、loss3 是否替换为相应有意义的命名？

tianxin1860 reviewed (commented):
Leave some comments

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| ----- | ----- | ------- | ----- | ----- | ----- | ----- | ---------- |
+| 74.28 | 57.33 | 61.72   | 81.06 | 76.20 | 86.51 | 78.77 | 73.70      |
+
+### 你可以这样导出Fine-tuning之后的模型直接用于部署
```
这个作为标题不太合适，标题要尽可能简洁

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+# PP-MiniLM中文特色小模型
+
+
+PP-MiniLM中文特色小模型，模型结构同ERNIE，目前本案例主要包含六层transformer layer的模型的通用蒸馏，以及借助PaddleSlim对模型进行裁剪和量化，进一步提升推理速度。
```
我们自称为特色是否合适还请再看下。

这里的本案例可能也不太合适，现在是本模型或者本方案了，是自己的模型了

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+### 原理介绍
+
+PP-MiniLM模型的蒸馏方法介绍：
```
是否也提下体现下原版MiniLM，也能显出我们命名的由来

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+执行完成后，模型保存的路径位于`ofa_models/CLUEWSC2020/0.75/best_model/`
+
+### 导出裁剪后的模型：
```
标题里加: 好像也不太合适

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+### 环境要求
+
+本实验如果基于NVIDIA V100 32G 8卡进行，训练周期约为2-3天。若资源有限，可以直接下载这一步得到的模型跳过此步骤。
```
这个感觉不能算是环境要求

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+
+PP-MiniLM中文特色小模型，模型结构同ERNIE，目前本案例主要包含六层transformer layer的模型的通用蒸馏，以及借助PaddleSlim对模型进行裁剪和量化，进一步提升推理速度。
+
+PP-MiniLM融合了蒸馏、裁剪、量化、高性能推理技术，拥有精度高、推理速度快、参数规模小的特点：
```
这句和上面那句可以合起来

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+| 腾讯 RoBERTa 6l-768d | 59.7M         |        | 1.00x   | 69.74 | 66.36 | 59.95   | 77.00 | 71.39 | 71.05 | 82.83 | 71.19      |
+| PP-MiniLM 6l-768d    | 59.7M         |        | 1.00x   | 74.28 | 57.33 | 61.72   | 81.06 | 76.2  | 86.51 | 78.77 | 73.70      |
+| PP-MiniLM裁剪后      | 49.1M (+裁剪) |        | 1.15x   | 73.82 | 57.33 | 61.60   | 81.38 | 76.20 | 85.52 | 79.00 | 73.55      |
+| PP-MiniLM量化后      | 49.2M(+量化)  |        | 2.18x   | 73.61 | 57.18 | 61.49   | 81.26 | 76.31 | 84.54 | 77.67 | 73.15      |
```
量化后模型比上一步还更大了吗

guoshengCS commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,297 @@
+- `num_relation_heads` relation heads的个数，一般对于large size的教师模型是64，对于base size的教师模型是48。
+- `teacher_model_type`指示了教师模型类型，当前仅支持'ernie'、'roberta'。
+- `teacher_layer_index`蒸馏时使用的教师模型的层数
+- `student_layer_index` 蒸馏时使用的学生模型的层数
```
是想表示选用第几层吧，层数感觉可能会带来些误解

LiuChiachi committed:
# This is a combination of 2 commits.
# The first commit's message is:

update inference

# This is the 2nd commit message:

update

LiuChiachi committed:
solve conflicts

fix infer perf

remove useless comments

LiuChiachi committed:
update readme

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-ppminilm

LiuChiachi committed:
update readme

update readme

LiuChiachi committed:
delete useless char

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,305 @@
+# PP-MiniLM中文小模型
+
+PP-MiniLM 中文特小模型案例旨在提供训推一体的高精度、高性能小模型解决方案。
```
PP-MiniLM 中文特小模型案例旨在提供训推一体的高精度、高性能小模型及解决方案。

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,305 @@
+
+当前解决方案依托业界领先的 Task Agnostic 模型蒸馏技术、裁剪技术、量化技术，使得小模型兼具推理速度快、模型效果好、参数规模小的 3 大特点。
+
+- 推理速度快：我们集成了 PaddleSlim 的裁剪、量化技术进一步对小模型进行压缩，保证模型推理速度达到原先的2.18倍；
```
推理速度快:  依托 PaddleSlim 的裁剪、量化技术进一步小模型进行压缩, 使得 PP-MiniLM 量化模型 GPU 推理速度相比 Bert-base 加速比高达 3.56；

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,305 @@
+
+- 推理速度快：我们集成了 PaddleSlim 的裁剪、量化技术进一步对小模型进行压缩，保证模型推理速度达到原先的2.18倍；
+
+- 精度高： 我们以 MiniLMv2 提出的 Multi-Head Self-Attention Relation Distillation 技术为基础，通过引入样本间关系知识蒸馏做了进一步算法优化。我们的6层、hidden size为768的模型，在CLUE上的平均准确率分别高于TinyBERT、UER-py RoBERTa同样大小的模型2.66%、1.51%。
```
精度高: 我们以 MiniLMv2 提出的 Multi-Head Self-Attention Relation Distillation 技术为基础，通过引入样本间关系知识蒸馏做了进一步算法优化,  6 层 PP-MiniLM 模型在 CLUE 数据集上比 12 层 Bert-base-chinese 高 0.23%，比同等规模的 TinyBERT、UER-py RoBERTa 分别高 2.66%、1.51%；

tianxin1860 commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,305 @@
+
+- 精度高： 我们以 MiniLMv2 提出的 Multi-Head Self-Attention Relation Distillation 技术为基础，通过引入样本间关系知识蒸馏做了进一步算法优化。我们的6层、hidden size为768的模型，在CLUE上的平均准确率分别高于TinyBERT、UER-py RoBERTa同样大小的模型2.66%、1.51%。
+
+- 参数规模小：依托 PaddleSlim 裁剪技术，在精度几乎无损(-0.15)条件下将模型宽度压缩 1/4。
```
参数规模小：依托 PaddleSlim 裁剪技术，在精度几乎无损(-0.15%)条件下将模型隐层宽度压缩 1/4，模型参数量减少 28%；

LiuChiachi replied:
已经修改，十分感谢～

tianxin1860 reviewed (commented):
Leave some comments

LiuChiachi committed:
update reamde

update code and readme

update readme

Add serial number to readme

update readme

Added a catalog

fix a catalog bug

fix a catalog bug

LiuChiachi committed:
update reamde

update reamde

update readme

update reamde

update readme

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,433 @@
+
+| Model                   | #Params | #FLOPs | Speedup | AFQMC | TNEWS | IFLYTEK | CMNLI | OCNLI | WSC   | CSL   | CLUE平均值 |
+| ----------------------- | ------- | ------ | ------- | ----- | ----- | ------- | ----- | ----- | ----- | ----- | ---------- |
+| Bert<sub>base</sub>     | 102.3M  | 10.87B | 1.00x   | 74.17 | 57.17 | 61.14   | 81.14 | 75.08 | 80.26 | 81.47 | 72.92      |
```
BERT，作为模型名的BERT统一大写

LiuChiachi replied:
感谢，已经修改

ZeyuChen commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,433 @@
+
+### 环境说明
+
+本实验基于NVIDIA Tesla V100 32G 8卡进行，训练周期约为2-3天。若资源有限，可以直接[下载PP-MiniLM(6L768H)](https://bj.bcebos.com/paddlenlp/models/transformers/ppminilm/6l-768h)用于下游任务的微调。
```
需要手动下载吗？是否可以告诉大家用`from_pretrained`接口自动下载？

LiuChiachi replied:
不需要，已经加上了用`from_pretrained`导入的示例，并在`modeling.py`和`tokenizer.py`加上ppminilm相关配置

jiweibo commented on examples/model_compression/PP-MiniLM/inference/infer.py:
```diff
@@ -0,0 +1,235 @@
+        elif args.device == "xpu":
+            # set XPU configs accordingly
+            config.enable_xpu(100)
+        config.switch_use_feed_fetch_ops(False)  # could be deleted
```
1.8遗留的接口，2.0后可删去，建议去掉，不影响性能、正确性等

LiuChiachi replied:
感谢提示！已经删去。

jiweibo commented on examples/model_compression/PP-MiniLM/inference/infer.py:
```diff
@@ -0,0 +1,235 @@
+                    use_calib_mode=False)
+            print("Enable TensorRT is: {}".format(
+                config.tensorrt_engine_enabled()))
+            if args.collect_shape:
```
collect_shape目前在`if args.use_trt ` block内，但从概念上讲，这俩是并列的，建议同级处理

LiuChiachi replied:
感谢提示，非常重要～已经修改～

jiweibo commented on examples/model_compression/PP-MiniLM/inference/infer_perf.py:
```diff
@@ -0,0 +1,208 @@
+    return args
+
+
+class Predictor(object):
```
似乎大部分代码重复了，加个基类处理下？

LiuChiachi replied:
感谢提示，已经重新加了一个函数做性能测试。

jiweibo commented on examples/model_compression/PP-MiniLM/inference/infer_all.sh:
```diff
@@ -0,0 +1,12 @@
+for task in afqmc tnews iflytek cmnli ocnli cluewsc2020 csl
```
是否需要加copyright声明呢

LiuChiachi replied:
感谢提示，已经把shell上都加了copyright

LiuChiachi committed:
update readme, add general readme, remove 'ofa'

LiuChiachi committed:
remove infe_perf

jiweibo reviewed (dismissed):
LGTM for inference api

jiweibo commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,379 @@
+
+#### 环境要求
+
+这一步依赖安装有预测库的 PaddlePaddle 2.2.1。可以在[PaddlePaddle 官网](https://paddleinference.paddlepaddle.org.cn/user_guides/download_lib.html)根据机器环境选择合适的 Python 预测库进行安装。
```
这个url直达python？
https://paddleinference.paddlepaddle.org.cn/user_guides/download_lib.html#python

LiuChiachi replied:
感谢提示，已经更新

Commit:
Update README

Commit:
Update README

LiuChiachi committed:
remove blank space between Chinese characters and numbers

LiuChiachi committed:
solve conflicts

LiuChiachi committed:
add blank space

fix director

chenxiaozeng commented on examples/model_compression/PP-MiniLM/README.md:
```diff
@@ -0,0 +1,378 @@
+
+## 导入 PP-MiniLM
+
+PP-MiniLM是使用任务无关蒸馏方法，以 `roberta-wwm-ext-large` 做教师模型蒸馏产出的包含 6 层 Transformer Encoder Layer、Hidden Size 为 768 的中文预训练小模型，在[中文任务测评基准 CLUE](https://github.com/CLUEbenchmark/CLUE) 上七个分类任务上的模型精度超过 BERT<sub>base</sub>、TinyBERT<sub>6</sub>、UER-py RoBERTa L6-H768、RBT6。
```
roberta-wwm-ext-large 作为教师模型，6层ERNIE作为学生模型是吧？感觉体现下ERNIE比较清晰

LiuChiachi replied:
感谢建议，已经体现 6 层 ERNIE了。

tianxin1860 reviewed (dismissed):
LGTM

LiuChiachi committed:
update data in readme

LiuChiachi committed:
add blank space

LiuChiachi committed:
update readme

LiuChiachi committed:
Merge branch 'develop' of https://github.com/PaddlePaddle/PaddleNLP into add-ppminilm

ZeyuChen reviewed (approved):
LGTM

LiuChiachi commented:
Thank you all:)🙏
